### PR TITLE
Support cropping and trapezoidal mask

### PR DESCRIPTION
**Pull request by DocGarbanzo**
# Added Cropping and Trapezoidal mask 

## Missing features in DC 4.x
Cropping was already supported in DC 3.x. and dropped off when we moved to 4.x. This is now back. Both cropping and trapezoidal mask are going using the augmentations class, but are enabled with a different configuration called `TRANSFORMATION`.  If transformations are set in training then they have to be enabled on the car, too. Cropping / trapezoidal transformations can be mixed with multiply or blur augmentations which only affects the training (but not the validation).

## Support in the App
The UI now also supports cropping / trapezoidal mask. Note, if this is intended to be used in training, then these have to be explicitly set in the configuration through the text input. Here is what these look like:

* Trapezoidal: This removes parts of the image, i.e. converts the pixels to black. There no change in the image dimension.
Screenshot 2021-07-17 at 19 35 53

* Cropping: This crops left, right, bottom, top part of the image and rescales the cropped image back to the original size. Therefore there is no image dimension change either! Here we cropped 45 pixels from the top.
Screenshot 2021-07-17 at 19 36 14

* Trapezoidal w/ augmentation and blur: You can combine the different augmentations like here.
Screenshot 2021-07-17 at 19 36 58

## Related
* Updated configs and car templates so crop / trapezoidal can be set
* Added cropping to test and checked convergence

### CODE REVIEW

**Review by Ezward — DISMISSED** [review]
I'm unable to test this pull request because it is missing install/envs/ubuntu.  Maybe you need to rebase against dev?

**Comment by Ezward** [comment]
PS: thank you for making these changes.  Cropping and trapezoidal masking are really important features for getting a model to work in the real world.

File: donkeycar/pipeline/training.py [hunk 1]
```diff
@@ -117,8 +117,11 @@ def train(cfg: Config, tub_paths: str, model: str = None,
     print(f'Records # Training {len(training_records)}')
     print(f'Records # Validation {len(validation_records)}')
 
-    training_pipe = BatchSequence(kl, cfg, training_records, is_train=True)
-    validation_pipe = BatchSequence(kl, cfg, validation_records, is_train=False)
+    # We need augmentation in validation when using crop / trapeze
+    apply_aug = getattr(cfg, 'APPLY_AUGMENTATION_IN_VALIDATION', True)
+    training_pipe = BatchSequence(kl, cfg, training_records, apply_aug=True)
+    validation_pipe = BatchSequence(kl, cfg, validation_records,
```
[the remark this answers is not in the document]
Comment (DocGarbanzo) [reply donkeycar/pipeline/training.py:123]: Yes, it makes sense to separate them. I will do that.

**Comment by DocGarbanzo** [comment]
> I'm unable to test this pull request because it is missing install/envs/ubuntu. Maybe you need to rebase against dev?

There are no changes to `ubuntu.yml` - where do you see a problem?

**Review by tikurahul — APPROVED** [review]
🏆

**Comment by DocGarbanzo** [comment]
@Ezward , @tikurahul - I made the changes of dispatching between transformations and augmentations, rebased everything and squeezed into a single clean commit. Can you have a quick look again, pls, then I can merge.

File: donkeycar/pipeline/training.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DocGarbanzo) [reply donkeycar/pipeline/training.py]: This is done.

**Pull request merged**